builtin/logical/database/backend.go

```diff
@@ -286,3 +485,327 @@ cassandra, mssql, mysql, postgres
+//
+// This method loops through the priority queue, popping the highest priority
+// item until it encounters the first item that does not yet need rotation,
+// based on the current time.
+func (b *databaseBackend) rotateCredentials(ctx context.Context, s logical.Storage) error {
+	for {
+		item, err := b.credRotationQueue.Pop()
+		if err != nil {
+			if err == queue.ErrEmpty {
+				return nil
+			}
+			return err
+		}
+
+		// validate the role still exists
+		role, err := b.StaticRole(ctx, s, item.Key)
+		if err != nil {
+			b.logger.Warn("unable load role", "role", item.Key, "error", err)
+			continue
+		}
+		if role == nil {
+			b.logger.Warn("role not found", "role", item.Key, "error", err)
+			continue
+		}
+
+		if time.Now().Unix() > item.Priority {
+			// We've found our first item not in need of rotation
+			input := &setStaticAccountInput{
+				RoleName: item.Key,
+				Role:     role,
+			}
+
+			// If there is a WAL entry related to this Role, the corresponding WAL ID
+			// should be stored in the Item's Value field.
+			if walID, ok := item.Value.(string); ok {
+				walEntry := b.findStaticWAL(ctx, s, walID)
+				if walEntry != nil && walEntry.NewPassword != "" {
+					input.Password = walEntry.NewPassword
+					input.WALID = walID
+				}
+			}
+
+			resp, err := b.setStaticAccount(ctx, s, input)
+			if err != nil {
+				b.logger.Warn("unable to rotate credentials in periodic function", "error", err)
+				// update the priority to re-try this rotation and re-add the item to
+				// the queue
+				item.Priority = time.Now().Add(10 * time.Second).Unix()
+
+				// preserve the WALID if it was returned
+				if resp.WALID != "" {
+					item.Value = resp.WALID
+				}
+
+				if err := b.pushItem(item); err != nil {
+					b.logger.Warn("unable to push item on to queue", "error", err)
+				}
+				// go to next item
+				continue
+			}
+
+			lvr := resp.RotationTime
+			if lvr.IsZero() {
+				lvr = time.Now()
+			}
+
+			// update priority and push updated Item to the queue
+			nextRotation := lvr.Add(role.StaticAccount.RotationPeriod)
+			item.Priority = nextRotation.Unix()
+			if err := b.pushItem(item); err != nil {
+				b.logger.Warn("unable to push item on to queue", "error", err)
+			}
+		} else {
+			// highest priority item does not need rotation, so we push it back on
+			// the queue and break the loop
+			b.pushItem(item)
+			break
+		}
+	}
+	return nil
+}
+
+// findStaticWAL loads a WAL entry by ID. If found, only return the WAL if it
+// is of type staticWALKey, otherwise return nil
+func (b *databaseBackend) findStaticWAL(ctx context.Context, s logical.Storage, id string) *setCredentialsWAL {
+	wal, err := framework.GetWAL(ctx, s, id)
+	if err != nil {
+		b.Logger().Warn("error reading WAL", "id", id, "error", err)
+		return nil
+	}
+
+	if wal == nil || wal.Kind != staticWALKey {
+		return nil
+	}
+
+	var walEntry setCredentialsWAL
+	if mapErr := mapstructure.Decode(wal.Data, &walEntry); err != nil {
```

Comment:
```suggestion
	if mapErr := mapstructure.Decode(wal.Data, &walEntry); mapErr != nil {
```

Reply:
Done in c0907c0aa

Reply:
well, similar: I changed it to just use `err`